src/main/java/net/snowflake/client/core/HeartbeatBackground.java

```diff
@@ -43,7 +43,7 @@
    * for it. This is to take care of the case when some application does not close session before it
    * goes out of scope.
    */
-  WeakHashMap<SFSession, Boolean> sessions = new WeakHashMap<>();
+  WeakHashMap<SFSessionImpl, Boolean> sessions = new WeakHashMap<>();
```

Comment:
A quick question. why the implementation class is exposed while the interface exists for abstraction?

Reply:
Good question. There's a bunch of specialized code (e.g., `HeartbeatBackground`) that we don't expect will be needed in our 'other' implementation of the JDBC interface. Therefore many classes apply only to the `...Impl` default implementations, as they will call upon methods defined only in the default client implementation.

We can sync further offline if you have more questions.

Reply:
ok, but can you please elaborate what is the value of enforcing the use of `...Impl` over the interface `SFSession`? In my reading, using the interface will reduce the code changes as well as add a flexibility of adding other implementation in the future? Or you don't want those places to use the other implementation at all?

Reply:
Exactly, in some places we don't want to use the other implementation at all. For example, `SFStatementImpl` will call `session.renewSession()`. In our other implementation, we don't expect to need to implement that at all, as that will be handled elsewhere.

I think you may be missing some context; I'm happy to go into the background deeper offline.

Reply:
ok, I'm ok with either case. This is an implementation details. I'll leave further discussion to @sfc-gh-mknister if she has concern.